lightning/restore/restore.go

```diff
@@ -529,38 +597,50 @@ func (t *TableRestore) restoreTable(
 				defer wg.Done()
 				tag := fmt.Sprintf("%s:%d", t.tableName, eid)
 
-				closedEngine, closedEngineWorker, err := t.restoreEngine(ctx, rc, eid, ecp)
+				dataClosedEngine, dataWorker, err := t.restoreEngine(ctx, rc, indexEngine, eid, ecp)
 				rc.tableWorkers.Recycle(w)
 				if err != nil {
 					engineErr.Set(tag, err)
 					return
 				}
 
-				defer rc.closedEngineLimit.Recycle(closedEngineWorker)
-				if err := t.importEngine(ctx, closedEngine, rc, eid, ecp); err != nil {
+				defer rc.closedEngineLimit.Recycle(dataWorker)
+				if err := t.importEngine(ctx, dataClosedEngine, rc, eid, ecp); err != nil {
 					engineErr.Set(tag, err)
 				}
-			}(restoreWorker, engineID, engine)
+			}(restoreWorker, engine.EngineID, engine)
 		}
 
 		wg.Wait()
 
 		common.AppLogger.Infof("[%s] import whole table takes %v", t.tableName, time.Since(timer))
 		err := engineErr.Get()
-		rc.saveStatusCheckpoint(t.tableName, -1, err, CheckpointStatusImported)
 		if err != nil {
 			return errors.Trace(err)
 		}
+
+		// If index engine file has been closed but not imported only if context cancel occurred
+		// when `importKV()` execution, so `UnsafeCloseEngine` and continue import it.
+		if indexEngineCp.Status == CheckpointStatusClosed {
```

Comment:
what if lightning execute L793 (L794 doesn't take effect), and then lightning was killed directly?

Reply:
addressed